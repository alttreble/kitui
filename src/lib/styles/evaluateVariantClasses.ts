import type { Variants } from '$lib';
import { twMerge } from 'tailwind-merge';

export default function evaluateVariantClasses<T, U extends Record<string, string>>(props: T, variants: Variants<T, U>) {
	const applicableStyles = {} as Record<keyof U, string>;
	for (const variant of variants) {
		let currentVariantApplicable = true;
		for (const propKey of Object.keys(variant.props) as (keyof T)[]) {
				if (props[propKey] !== variant.props[propKey]) {
					currentVariantApplicable = false;
					break;
				}
		}
		if (!currentVariantApplicable) continue;

        // Evaluate all applicable styles
        for (const styleKey of Object.keys(variant.styles) as (keyof U)[]) {
            if (applicableStyles[styleKey]) {
                applicableStyles[styleKey] += ' ' + variant.styles[styleKey];
            } else {
                applicableStyles[styleKey] = variant.styles[styleKey];
            }
        }
    }
    // Apply tailwind-merge to all styles
    for (const styleKey of Object.keys(applicableStyles) as (keyof U)[]) {
        applicableStyles[styleKey] = twMerge(applicableStyles[styleKey]);
    }

    return applicableStyles;
}
