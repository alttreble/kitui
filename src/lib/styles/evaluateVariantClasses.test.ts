import evaluateVariantClasses from '$lib/styles/evaluateVariantClasses';
import type { ButtonProps, Variants } from '$lib';
import { expect } from 'vitest';

it("should get styles for primary filled button", () => {
	const props: ButtonProps = {
		element: "button",
		type: "filled",
		color: "primary"
	};

	const variants: Variants<ButtonProps> = [
		{
			props: {},
			class: "h-fit w-fit rounded-lg py-2 px-4"
		},
		{
			props: { type: "outlined" },
			class: "border transition-colors font active:bg-gray-100/70",
		},
		{
			props: { type: "outlined", color: "primary" },
			class: `text-primary border-outline
									shadow-sm shadow-primary-200/50
         					hover:border-primary-500`
		},
		{
			props: {type: "filled"},
			class: "text-white hover:opacity-90 transition-opacity"
		},
		{
			props: {type: "filled", color: "primary"},
			class: `bg-gradient-to-b from-primary-30 to-primary
         					shadow-md shadow-primary-200/50
         					active:from-primary-600 active:to-primary-700`
		},
		{
			props: {type: "filled", color: "secondary"},
			class: `bg-gradient-to-b from-secondary-30 to-secondary-40
									shadow-md shadow-secondary-200/50
									active:from-secondary-600 active:to-secondary-700`
		}
	]

	const evaluatedClasses = evaluateVariantClasses(props, variants)
	expect(evaluatedClasses).toMatchSnapshot();
});
