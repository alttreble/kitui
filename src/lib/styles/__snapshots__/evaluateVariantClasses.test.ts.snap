// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should get styles for primary filled button 1`] = `"h-fit w-fit rounded-lg py-2 px-4 text-white hover:opacity-90 transition-opacity bg-gradient-to-b from-primary-30 to-primary shadow-md shadow-primary-200/50 active:from-primary-600 active:to-primary-700"`;
